src/lib/transition_handler/validator.ml

```diff
@@ -1,149 +1,122 @@
-      match
-        log_assert
-          (Consensus.Mechanism.is_valid
-             (consensus_state transition)
-             ~time_received)
-          "failed consensus validation"
-        && log_assert
-             ( Consensus.Mechanism.select ~logger
-                 ~existing:(consensus_state_verified root)
-                 ~candidate:(consensus_state transition)
-                 ~time_received
-             = `Take )
-             "was not better than transition frontier root"
-      with
-      | `Valid ->
-          (* TODO: what conditions should we punish? *)
-          (* TODO:
-      let length = External_transition.protocol_state t |> Protocol_state.blockchain_state |> Blockchain_state.length in
-      log_assert
-        (match Root_history.find (Transition_frontier.root_history frontier) (length - k) with
-        | `Known h -> State_hash.equal h (External_transition.frontier_root_hash t)
-        | `Unknown -> true
-        | `Out_of_bounds ->
-          Logger.info logger "expected root of transition was out of bounds";
-          false)
-        "transition frontier root hash was invalid"
-      *)
-          let%map proof_is_valid =
-            State_proof.verify
-              (External_transition.protocol_state_proof transition)
-              (External_transition.protocol_state transition)
-          in
-          log_assert proof_is_valid "proof was invalid"
-      | x -> return x
-
-  let verify_transition ~staged_ledger ~transition :
-      External_transition.Verified.t Or_error.t Deferred.t =
-    let open Deferred.Or_error.Let_syntax in
-    let diff = External_transition.staged_ledger_diff transition in
-    let%bind verified_diff =
-      Staged_ledger.verified_diff_of_diff staged_ledger diff
-    in
-    return
-      (External_transition.Verified.create
-         ~protocol_state:(External_transition.protocol_state transition)
-         ~protocol_state_proof:
-           (External_transition.protocol_state_proof transition)
-         ~staged_ledger_diff:verified_diff)
+      let rec check_cases = function
+        | [] -> return `Valid
+        | (cond_lazy, failure_reason) :: tail ->
+            if%bind Lazy.force cond_lazy then check_cases tail
+            else return (`Invalid failure_reason)
+      in
+      let validation_cases =
+        [ ( lazy
+              (return
+                 ( `Take
+                 = Consensus.Mechanism.select ~logger
+                     ~existing:(consensus_state root_protocol_state)
+                     ~candidate:(consensus_state protocol_state) ))
+          , "consensus state was not selected over transition frontier root \
+             consensus state" )
+        ; ( lazy
+              (State_proof.verify
+                 (External_transition.protocol_state_proof transition)
+                 protocol_state)
+          , "protocol state proof was not valid" ) ]
+      in
+      let validation_cases =
+        Option.value ~default:validation_cases
```

Comment:
Pet peeve: reusing variable names, which leads to potential confusion. Could one of the `validation_cases` be renamed?

Reply:
This pattern is typical in our code as of right now. There are no rules in the style guide one way or the other though in reference to this form of variable shadowing. I'm personally in favor of this pattern, but we could still change the rules. Do we have a process around proposing additions to our style guide? @bkase?